File: computing/interview/data-structures/cpp/hash_map.cpp
// Hash Map — using std::unordered_map (O(1) avg) and std::map (O(log n), sorted)
#include <iostream>
#include <string>
#include <unordered_map>
#include <map>
#include <vector>

// Classic use-case: two-sum
std::vector<int> twoSum(const std::vector<int>& nums, int target) {
    std::unordered_map<int, int> seen; // value -> index
    for (int i = 0; i < static_cast<int>(nums.size()); ++i) {
        int complement = target - nums[i];
        if (seen.count(complement)) return {seen[complement], i};
        seen[nums[i]] = i;
    }
    return {};
}

int main() {
    // --- unordered_map: O(1) average ---
    std::unordered_map<std::string, int> freq;
    std::vector<std::string> words = {"apple", "banana", "apple", "cherry", "banana", "apple"};
    for (const auto& w : words) ++freq[w];

    std::cout << "Word frequencies:\n";
    for (const auto& [word, count] : freq)
        std::cout << "  " << word << ": " << count << "\n";

    // Lookup
    std::cout << "Contains 'apple': " << freq.count("apple") << "\n";
    freq.erase("banana");
    std::cout << "After erase 'banana', size: " << freq.size() << "\n";

    // --- std::map: O(log n), keys are sorted ---
    std::map<int, std::string> sorted_map = {{3, "c"}, {1, "a"}, {2, "b"}};
    std::cout << "\nSorted map (keys in order):\n";
    for (const auto& [k, v] : sorted_map)
        std::cout << "  " << k << " -> " << v << "\n";

    // --- Two-sum demo ---
    std::vector<int> nums = {2, 7, 11, 15};
    auto result = twoSum(nums, 9);
    std::cout << "\nTwo-sum indices: [" << result[0] << ", " << result[1] << "]\n";

    return 0;
}

File: computing/interview/data-structures/python/hash_map.py
# Hash Map — Python dict (built-in, O(1) avg) and collections.Counter / defaultdict
from collections import Counter, defaultdict


def two_sum(nums: list[int], target: int) -> list[int]:
    """Classic use-case: two-sum."""
    seen: dict[int, int] = {}   # value -> index
    for i, x in enumerate(nums):
        complement = target - x
        if complement in seen:
            return [seen[complement], i]
        seen[x] = i
    return []


# --- Plain dict ---
freq: dict[str, int] = {}
words = ["apple", "banana", "apple", "cherry", "banana", "apple"]
for w in words:
    freq[w] = freq.get(w, 0) + 1       # manual frequency count

print("Frequencies:", freq)
print("Contains 'apple':", "apple" in freq)
del freq["banana"]
print("After deleting 'banana':", freq)

# --- defaultdict: no KeyError on missing keys ---
graph: defaultdict[str, list[str]] = defaultdict(list)
graph["a"].append("b")
graph["a"].append("c")
graph["b"].append("d")
print("\nDefault-dict graph:", dict(graph))

# --- Counter: frequency counter with extras ---
counter = Counter(["a", "b", "a", "c", "b", "a"])
print("\nCounter:", counter)
print("Most common 2:", counter.most_common(2))
counter.update(["b", "b"])
print("After update:", counter)

# --- Two-sum demo ---
print("\nTwo-sum([2,7,11,15], 9):", two_sum([2, 7, 11, 15], 9))
